source/validate.cpp

```diff
@@ -448,22 +450,38 @@ spv_result_t ModuleLayoutPass(ValidationState_t& _,
   if (_.is_enabled(SPV_VALIDATE_LAYOUT_BIT)) {
     SpvOp opcode = inst->opcode;
 
-    if (libspirv::ModuleLayoutSection::kModule == _.getLayoutStage()) {
+    if (_.getLayoutStage() < kLayoutFunctionDeclarations) {
       // Module scoped instructions are processed by determining if the opcode
       // is part of the current stage. If it is not then the next stage is
       // checked.
       while (_.isOpcodeInCurrentLayoutStage(opcode) == false) {
-        // TODO(umar): Check if the MemoryModel instruction has executed
         _.progressToNextLayoutStageOrder();
-        if (_.getLayoutStage() == libspirv::ModuleLayoutSection::kFunction) {
+
+        if (_.getLayoutStage() == kLayoutMemoryModel &&
```

Comment:
Why single out this stage to check the ordering?  Doesn't the same principle apply to every stage?  Eg, `OpMemoryModel` before `OpCapability` or `OpString` before `OpExecutionMode`.